tools/mason/MasonPublish.chpl

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+use MasonUtils;
+use Spawn;
+use FileSystem;
+use TOML;
+use MasonEnv;
+use MasonNew;
+use MasonModify;
+
+
+proc masonPublish(args) throws {
+  try! {
+    if args.size == 2 {
+      masonPublishHelp();
+      exit(0);
+    }
+    var dry = false;
+    var username = "";
+    if args.size > 2 {
+      for arg in args[2..] {
+        if arg == '-h' || arg == '--help'{
+          masonPublishHelp();
+          exit(0);
+        }
+        else if arg == "--dry-run" {
+          dry = true;
+        }
+        else {
+          username = arg;
+        }
+      }
+      if doesGitOriginExist() {
+        if dry {
+          dryRun(username);
+        }
+        else {
+          publishPackage(username);
+        }
+      }
+      else {
+            throw new owned MasonError('Must have package set up as git repo to publish');
+      }
+    }
+  }
+  catch e: MasonError {
+    writeln(e.message());
+    exit(1);
+  }
+}
+
+
+proc dryRun(username: string) throws {
+  var fork = false;
+  var checkIfForkExists = ('git ls-remote https://github.com/' + username + '/mason-registry').split();
+  var p = spawn(checkIfForkExists, stdout=PIPE);
+  p.wait();
+  if p.exit_status == 0 {
+    fork = true;
+  }
+  var git = false;
+  if doesGitOriginExist() {
+    git = true;
+  }
+  if git && fork {
+    writeln('Package can be published to the mason-registry');
+  }
+  else {
+    if fork == false {
+      throw new owned MasonError('mason-registry is not forked on your GitHub');
+    }
+    else {
+      throw new owned MasonError('Package does not gave a git origin');
+    }
+  }
+} 
+
+
+proc publishPackage(username: string) throws {
+  forkMasonReg(username);
+  branchMasonReg(username);
+  const cwd = getEnv("PWD");
+  const package=  addPackageToBricks();
+  here.chdir(cwd + "/mason-registry");
+  const name = getName();
+  const url = gitUrl();
+  const projectHome = getProjectHome(cwd);
+  here.chdir(cwd +"/mason-registry/");
+  runCommand("git add .");
```

Comment:
While we don't have a [libgit2 interface](https://libgit2.org/libgit2/#HEAD) in Chapel, it could be cool to wrap some of these git calls into a library-like interface. This could allow us to add some special case error checking for some commands based on the `runCommand` output. Bonus points if these calls are used in other mason modules. 

I wouldn't expect anything comprehensive, but just wrapping the calls we need for mason purposes e.g. the following commands could be:

```chpl
// Just winging it with the mock interface here:
gitAdd('.');
gitCommit(msg=package);
gitPush(package, setUpstream='origin');
gitRemoteAdd('upstream', 'https://github.com/chapel-lang/mason-registry');
gitRequestPull('upstream/master', 'https://github.com/chapel-lang/mason-registry');
```

This is totally optional for this PR, but we should definitely consider it in the future if you decide not to pursue it here.

Reply:
I think this should be a separate issue as there are some git commands in a couple different Mason functions so it probably makes sense to tackle those all at once.